src/liballoc/collections/vec_deque.rs

```diff
@@ -62,6 +62,20 @@ impl<T: Clone> Clone for VecDeque<T> {
     fn clone(&self) -> VecDeque<T> {
         self.iter().cloned().collect()
     }
+
+    fn clone_from(&mut self, other: &Self) {
+        self.truncate(other.len());
+
+        let mut iter_other = other.iter();
+        for (elem, elem_other) in self.iter_mut().zip(&mut iter_other) {
+            elem.clone_from(elem_other);
+        }
```

Comment:
Instead of `zip` and such here, could this just be `truncate`, `reserve_exact`, the `extend` on the two parts from [`other.as_slices()`](https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.as_slices)?  Or maybe take advantage of an empty VecDeque being O(1) convertible to `Vec`, and re-use the `extend_from_slice` optimizations?  It would be nice for the memcpy specializations to happen here...

Reply:
That sounds enticing but it would lose the recursive use of `clone_from`?  Of course, that recursive use was not present before this PR either, so it's not mandatory.

Reply:
I've been working on such a thing. For replacing the zip part, you should be able to maintain the recursive use of clone_from (which I think is a nice feature), you just need to use clone_from_slice on up to 3 different slices (depending on overlap). However, it does seem a little wasteful to extend from the iterator, and so far my fix for that involves a lot of (unsafe) buffer_write calls and a match expression as long as the one in wrap_copy. Certainly any ideas about clean and efficient ways to do the extend portion would be very welcome.

Reply:
Good point, @bluss.  I guess there's an opportunity for improvement in Vec too:

https://github.com/rust-lang/rust/blob/master/src/liballoc/vec.rs#L2027-L2038

(Doesn't need to be this PR, of course.)